pygmt/src/grd2xyz.py

```diff
@@ -172,25 +167,37 @@ def grd2xyz(grid, output_type="pandas", outfile=None, **kwargs):
         # Reverse the dims because it is rows, columns ordered.
         dataframe_header = [grid.dims[1], grid.dims[0], grid.name]
 
-    with GMTTempFile() as tmpfile:
-        with Session() as lib:
-            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
-            with file_context as infile:
-                if outfile is None:
-                    outfile = tmpfile.name
-                lib.call_module(
-                    module="grd2xyz",
-                    args=build_arg_string(kwargs, infile=infile, outfile=outfile),
-                )
-
-        # Read temporary csv output to a pandas table
-        if outfile == tmpfile.name:  # if user did not set outfile, return pd.DataFrame
-            result = pd.read_csv(
-                tmpfile.name, sep="\t", names=dataframe_header, comment=">"
+    with Session() as lib:
+        with lib.virtualfile_from_data(
+            check_kind="raster", data=grid
+        ) as invfile, lib.virtualfile_to_gmtdataset() as outvfile:
```

Comment:
After wrapping `GMT_DATASET` and providing some data-conversion functions, now it's possible to get rid of temporary files from GMT modules that write table files. `grd2xyz` is used as an example.

`grd2xyz` (and other similar functions) supports three different output types: "file", "numpy" and "pandas". The simplest strategy is

1. Always write the table files into a `GMT_DATASET` object
2. If `output_type="file"`, then we can call the special `write` module to write it into a file
3. Otherwise, convert `GMT_DATASET` to a list of vectors and then convert vectors to numpy or pandas.

This strategy is implemented as option 1 in the codes. It's simple, so it's possible to wrap the if-tests (lines 180-186) in a separate function and reuse it in many places.

The downside is, that for `output_type="file"`, we need to one more GMT C API call to write the `GMT_DATASET` object to a file.

So, the option 2 is also implemented in the codes. If `output_type="file"`, we write the output directly into a file. It's more efficient, but the codes is less clear than those of option 1.

Reply:
Problem solved in commit f5a9d447d1b003f93207949ecf940b0ab8bceef4.